rust/lsp-lib/src/lsp_plugin.rs

```diff
@@ -0,0 +1,470 @@
+                    end: end_position,
+                }),
+                range_length: Some((end - start) as u64),
+                text: String::new(),
+            };
+
+            Ok(vec![text_document_content_change_event])
+        }
+        // Send the whole document again if it is not a trivial edit
+        else {
+            let text_document_content_change_event = TextDocumentContentChangeEvent {
+                range: None,
+                range_length: None,
+                text: view.get_document()?,
+            };
+
+            Ok(vec![text_document_content_change_event])
+        }
+    } else {
+        let text_document_content_change_event = TextDocumentContentChangeEvent {
+            range: None,
+            range_length: None,
+            text: view.get_document()?,
+        };
+
+        Ok(vec![text_document_content_change_event])
+    }
+}
+
+pub fn get_workspace_root_uri(
+    workspace_identifier: &String,
+    document_path: &Path,
+) -> Result<Url, Error> {
+    let identifier_os_str = OsStr::new(&workspace_identifier);
+
+    let mut current_path = document_path;
+    loop {
+        let parent_path = current_path.parent();
+        if let Some(path) = parent_path {
+            for entry in path.read_dir().expect("Cannot read directory contents") {
+                if let Ok(entry) = entry {
+                    if entry.file_name() == identifier_os_str {
+                        let path = entry.path();
+
+                        return Ok(Url::parse(
+                            format!("file://{}", path.to_str().unwrap()).as_ref(),
+                        )?);
+                    };
+                }
+            }
+
+            current_path = path;
+        } else {
+            break Err(Error::PathError);
+        }
+    }
+}
+
+/// Start a new Language Server Process by spawning a process given the parameters
+/// Returns a Arc to the Language Server Client which abstracts connection to the
+/// server
+fn start_new_server(
+    command: String,
+    arguments: Vec<String>,
+    file_extensions: Vec<String>,
+    workspace_identifier: Option<String>,
+    language_id: String,
+) -> Result<Arc<Mutex<LanguageServerClient>>, String> {
+    let mut process = Command::new(command)
+        .args(arguments)
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()
+        .expect("Error Occurred");
+
+    let writer = Box::new(BufWriter::new(process.stdin.take().unwrap()));
+
+    let language_server_client = Arc::new(Mutex::new(LanguageServerClient::new(
+        writer,
+        language_id,
+        file_extensions,
+        workspace_identifier,
+    )));
+
+    {
+        let ls_client = language_server_client.clone();
+        let mut stdout = process.stdout;
+
+        // Unwrap to indicate that we want thread to panic on failure
+        std::thread::Builder::new()
+            .name("STDIN-Looper".to_string())
```

Comment:
we're going to have duplicate names here; maybe use a name based on the command string?

Reply:
Indeed. Will do 👍